### PR TITLE
feat(contrib):  add 3D volume support to distance_transform

### PR DESCRIPTION
**Pull request by aarush-github**
## 📝 Description
Extends the distance_transform function and DistanceTransform module to support both 2D images and 3D volumes through a unified API. Changes:
- Add automatic dimension detection for 4D (B, C, H, W) and 5D (B, C, D, H, W) tensors
- Implement 3D Euclidean distance kernel using create_meshgrid3d and torch.norm
- Add comprehensive test suite for 3D functionality
- Update distance calculation to use consistent exponential formulation. The implementation uses cascaded convolutions to approximate Euclidean distance transforms efficiently for both 2D and 3D data. 

**Fixes** #3279

## 🛠️ Changes Made
- **Automatic dimension detection**: Function now detects whether input is 4D (2D images) or 5D (3D volumes)
- **3D kernel generation**: Implements 3D Euclidean distance kernel using `create_meshgrid3d` and `torch.norm`
- **Unified convolution interface**: Uses `filter2d` for 2D and `filter3d` for 3D seamlessly

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** Smoke tests, Kernel size variations, Value correctness test, Gradient checks for both 2D and 3D,  Loss gradient tests, Dynamo compilation test

- [x] **Manual Verification:** 
 **3D visualization**: Created slice-by-slice views of 3D distance volumes to verify proper distance propagation in all directions
**Comparative visualization**: Side-by-side comparison of kornia output vs SciPy EDT

- [x] **Performance/Edge Cases:** 

1.  **Empty inputs** (all zeros): Returns zeros correctly.
2.  **Full inputs** (all ones): Returns zeros correctly.
3.  **Boundary conditions**: Replicate padding ensures accurate edge distances.
4. **Numerical stability**: Explicit NaN/Inf handling, epsilon in log operation.
5.  **Non-square volumes**: Handles rectangular shapes properly.
6. **Invalid inputs**: Raises appropriate TypeErrors and ValueErrors.

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring, but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Comparison Image of Kornia vs Scipy:
Figure_5

### CODE REVIEW

File: kornia/contrib/distance_transform.py [hunk 1]
```diff
@@ -15,82 +15,105 @@
...
 
+from __future__ import annotations
+
 import math
 
 import torch
 from torch import nn
 
-from kornia.filters import filter2d
+from kornia.core import Tensor
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py]: check on main, as there was a massive refactor and core doesn't contain Tensor anymore, use directly torch.Tensor

same for utlils

**Review by edgarriba — CHANGES_REQUESTED** [review]
thanks! double check about the imports

File: kornia/contrib/distance_transform.py (hunk 1 above)
Comment (aarush-github) [reply kornia/contrib/distance_transform.py]: Yes, I noticed it when merging. Thanks for the review! :)

Comment (AdityaKumarSethia) [reply kornia/contrib/distance_transform.py]: Hey Aarush, I definitely don't want to step on your toes or take over your work here. I just saw the discussion about getting this unblocked and thought I'd give it a shot.

I managed to draft a version where the tests pass smoothly (fixed the typo too). You can check it out here: https://github.com/AdityaKumarSethia/kornia/tree/fix/issue-3279-distance-transform-3d

Feel free to use this or let me know if you want me to open the PR

Comment (AdityaKumarSethia) [reply kornia/contrib/distance_transform.py]: Check:
1. kornia/contrib/distance_transform.py
2. tests/contrib/test_conv_distance_transformer.py
Files

**Comment by aarush-github** [comment]
Hey @edgarriba, I have also added `test_noncontiguous_multi_channel` and updated the source code accordingly to enhance robustness. This can help prevent potential RuntimeError issues in user pipelines.

File: kornia/contrib/distance_transform.py (hunk 1 above)
Comment (aarush-github) [reply kornia/contrib/distance_transform.py]: Hey Aditya, thanks for sharing your repo. I’ve gone through both files, and I wanted to let you know that I’ve already fixed the issue you’re addressing. That said, please feel free to point out any discrepancies you notice. I’d be happy to take another look.

File: kornia/contrib/distance_transform.py [hunk 2]
```diff
@@ -15,82 +15,106 @@
...
+    if dim == 4:
+        n_iters = math.ceil(max(image.shape[2], image.shape[3]) / k_half)
+        grid = create_meshgrid(kernel_size, kernel_size, False, device, dtype)
+        grid = grid - k_half
 
-    # It is possible to avoid cloning the image if boundary = image, but this would require modifying the image tensor.
+        dist = torch.hypot(grid[0, ..., 0], grid[0, ..., 1])
+        kernel = torch.exp(-dist / h).unsqueeze(0)
+
+        def conv_op(x: torch.Tensor, k: torch.Tensor) -> torch.Tensor:
```
Comment (AdityaKumarSethia) [inline kornia/contrib/distance_transform.py]: Refactoring Suggestion (Style): I see conv_op is defined inside both conditional branches. While Python handles scope this way without crashing, it's generally safer and cleaner to assign the function alias explicitly.

Consider using direct assignment:
```Python
if dim == 4:
    conv_op = filter2d
else:
    # Bind the 3D-specific args once
    conv_op = filter3d
```

Comment (aarush-github) [reply kornia/contrib/distance_transform.py]: Hey Aditya, Thanks for the suggestion!
I personally like the original version (keeping 2D and 3D conv_op explicitly) because it is more readable and easily modifiable (if 3D block needs changes in the future). But @edgarriba let me know if you prefer what Aditya is suggesting. I will be happy to implement it. :)

File: kornia/contrib/distance_transform.py [hunk 3]
```diff
@@ -54,7 +52,7 @@ def distance_transform(image: Tensor, kernel_size: int = 3, h: float = 0.35) ->
         >>> dt = distance_transform(volume)
 
     """
-    if not isinstance(image, Tensor):
+    if not isinstance(image, torch.Tensor):
```
Comment (edgarriba) [inline kornia/contrib/distance_transform.py]: can you use the KORNIA_CHECK api to raise errors ?

File: kornia/contrib/distance_transform.py (hunk 2 above)
Comment (edgarriba) [reply kornia/contrib/distance_transform.py]: i do prefer also @aarush-github -- what;s more i'd have `_distance_transform_2d_impl` and `_distance_transform_3d_impl` dispatched by the if else branch

**Comment by aarush-github** [comment]
Hey @edgarriba, thank you for the review! I’ll implement those changes right away.
To confirm the plan:

1. Separate the logic into `_distance_transform_2d_impl` and `_distance_transform_3d_impl` functions.

2. Use the KORNIA_CHECK API (KORNIA_CHECK_IS_TENSOR and KORNIA_CHECK) for all input validation.

Let me know if I missed any other details you'd like to see! :)"

**Comment by edgarriba** [comment]
@aarush-github all good
